add basic support files for rtl87x2g series soc

YuzhuoLiuRTK opened the pull request:
In this PR, I add code for supporting the realtek rtl87x2g series soc.
With this PR, you can build and run a hello_world app with rtl8762gn_evb.
You can also use drivers such as urat, dma, gpio, adc and counter with rtl87x2g series soc.
Hi @RSGraceGuan @ZhiyuanTang17 , Please help to review this PR.

ZhiyuanTang17 committed:
dts: add dts support for rtl87x2g

ZhiyuanTang17 committed:
boards: rtl8762gn_evb: add rtl8762gn_evb support

ZhiyuanTang17 committed:
soc: rtk_bee: add rtl87x2g series soc support

ZhiyuanTang17 committed:
drivers: support clock-ctrl, pinctrl and uart of rtl87x2g

ZhiyuanTang17 committed:
soc: rtl87x2g: disable interrupt during vector table update

ZhiyuanTang17 committed:
drivers: clock_control: rename APBPeriph_RTZIP_CLOCK for rtl87x2g

YuzhuoLiuRTK committed:
drivers: cctl: add descriptions for rtl87x2g cctl binding file

Signed-off-by: liuyuzhuo <yuzhuo_liu@realsil.com.cn>

YuzhuoLiuRTK committed:
drivers: pinctrl: add descriptions for rlt87x2g pinctrl binding file

Signed-off-by: liuyuzhuo <yuzhuo_liu@realsil.com.cn>

YuzhuoLiuRTK committed:
drivers: uart: modify comments for rtl87x2g uart driver

Signed-off-by: liuyuzhuo <yuzhuo_liu@realsil.com.cn>

YuzhuoLiuRTK committed:
drivers: uart: rename CONFIG_RTK to CONFIG_REALTEK in kconfig file for rtl87x2g

Signed-off-by: liuyuzhuo <yuzhuo_liu@realsil.com.cn>

YuzhuoLiuRTK committed:
drivers: gpio: add gpio driver for rtl87x2g

YuzhuoLiuRTK committed:
tests: drivers: gpio_basic_api: add overlay for rtl8762gn_evb

YuzhuoLiuRTK committed:
drivers: dma: add dma driver for rtl87x2g

YuzhuoLiuRTK committed:
tests: drivers: dma: loop_transfer: add rtl8762gn_evb overlay and conf

YuzhuoLiuRTK committed:
tests: drivers: uart_async_api: add rtl8762gn_evb overlay

YuzhuoLiuRTK committed:
drivers: adc: add adc driver for rtl87x2g

YuzhuoLiuRTK committed:
tests: drivers: adc_api: add rtl8762gn_evb overlay

YuzhuoLiuRTK committed:
drivers: counter: add counter driver for rtl87x2g

YuzhuoLiuRTK committed:
tests: drivers: counter_basic_api: add rtl8762gn_evb overlay

YuzhuoLiuRTK committed:
Kconfig: rename RTK to REALTEK

YuzhuoLiuRTK committed:
drivers: pwm: add pwm driver for rtl87x2g

YuzhuoLiuRTK committed:
tests: drivers: pwm_api: add rtl8762gn_evb overlay

YuzhuoLiuRTK committed:
drivers: sensors: aon_qdec_rtl87x2g: add qdec driver for rtl87x2g

YuzhuoLiuRTK committed:
sampless: sensor: qdec: add rtl8762gn_evb overlay

YuzhuoLiuRTK committed:
dts: rename rtk to realtek

YuzhuoLiuRTK committed:
drivers: spi: add spi driver for rtl87x2g

YuzhuoLiuRTK committed:
tests: drivers: spi_loopback: add rtl8762gn_evb overlay and conf

YuzhuoLiuRTK committed:
drivers: kscan: add kscan driver for rtl87x2g

YuzhuoLiuRTK committed:
samples: drivers: kscan: add rtl8762gn_evb overlay

YuzhuoLiuRTK committed:
drivers: i2c: add i2c driver for rtl87x2g

YuzhuoLiuRTK committed:
drivers: gpio: add gpio debounce time configure for rtl87x2g

Add vendor flag "RTL87X2G_GPIO_INPUT_DEBOUNCE_MS" for dts and gpio_pin_configure api to config debounce time.

Signed-off-by: liuyuzhuo <yuzhuo_liu@realsil.com.cn>

YuzhuoLiuRTK committed:
drivers: uart: poll tx_fifo_len in poll_out api

Use tx_fifo_len < tx_fifo_size condition instead of tx_fifo_empty
condition to increase polling efficiency.

Signed-off-by: liuyuzhuo <yuzhuo_liu@realsil.com.cn>

YuzhuoLiuRTK committed:
drivers: kscan: add software debounce for rtl87x2g

Check scan result for specificed times to filter button press jitter.
Add all release interrupt to filter button release jitter.

Signed-off-by: liuyuzhuo <yuzhuo_liu@realsil.com.cn>

YuzhuoLiuRTK committed:
drivers: adc: add bypass mode and get voltage for rtl87x2g
Add properties "is-bypass-mode" in binding file and add releated code in
c file for users to configure bypass mode in dts. Call ADC_GetVoltage
api in adc_read api to convert raw sample data to actual voltage for
upper use.

Signed-off-by: liuyuzhuo <yuzhuo_liu@realsil.com.cn>

YuzhuoLiuRTK committed:
drivers: counter: modify prescaler value for rtl87x2g

Signed-off-by: liuyuzhuo <yuzhuo_liu@realsil.com.cn>

YuzhuoLiuRTK committed:
drivers: clock_control: rename 24badc for rtl87x2g

Signed-off-by: liuyuzhuo <yuzhuo_liu@realsil.com.cn>

YuzhuoLiuRTK committed:
drivers: gpio: fix flase trigger for rtl87x2g

add delay flow after configure gpio trigger mode to avoid false trigger

Signed-off-by: liuyuzhuo <yuzhuo_liu@realsil.com.cn>

YuzhuoLiuRTK committed:
drivers: adc: rename KConfig for rtl87x2g adc

Signed-off-by: liuyuzhuo <yuzhuo_liu@realsil.com.cn>

YuzhuoLiuRTK committed:
drivers: counter: rename KConfig for rtl87x2g tim

Signed-off-by: liuyuzhuo <yuzhuo_liu@realsil.com.cn>

YuzhuoLiuRTK committed:
drivers: sensor: rename KConfig for rtl87x2g aon_qdec

Signed-off-by: liuyuzhuo <yuzhuo_liu@realsil.com.cn>

YuzhuoLiuRTK committed:
drivers: pwm: rename KConfig for rtl87x2g pwm

Signed-off-by: liuyuzhuo <yuzhuo_liu@realsil.com.cn>

YuzhuoLiuRTK committed:
drivers: i2c: fix compiling warning for rlt87x2g i2c

Signed-off-by: liuyuzhuo <yuzhuo_liu@realsil.com.cn>

YuzhuoLiuRTK committed:
drivers: usb: add usb driver for rtl87x2g

Signed-off-by: liuyuzhuo <yuzhuo_liu@realsil.com.cn>

ZhiyuanTang17 commented on drivers/adc/Kconfig.rtl87x2g:
```diff
@@ -0,0 +1,16 @@
+	select REALTEK_ADC_ADC
+    select REALTEK_USING_SDK_LIB
+	help
+	  This option enables the adc driver for RTL87X2G family of
+	  processors.
+	  Say y if you wish to use adc on RTL87X2G MCU.
```
"RTL87X2G family of processors" seems inappropriate, " the rtl87x2g series SoC" may be better?
Do you mean "Set y", rather than "Say y"?

YuzhuoLiuRTK replied:
modified

YuzhuoLiuRTK committed:
drivers: modify KConfig for rtl87x2g drivers

Signed-off-by: liuyuzhuo <yuzhuo_liu@realsil.com.cn>

YuzhuoLiuRTK committed:
Merge branch 'rtkconnectivity:realtek-main-v3.5.0' into realtek-v3.5.0

ZhiyuanTang17 committed:
dts: update the devicetree of rtl8762gn_evb

ZhiyuanTang17 committed:
soc: realtek_bee: update the soc configuration

ZhiyuanTang17 committed:
boards: update default kconfig of rtl8762gn_evb

ZhiyuanTang17 committed:
drivers: wdt: support wdt driver for rtl87x2g series

YuzhuoLiuRTK committed:
drivers: modify KConfig for rtl87x2g drivers

Signed-off-by: liuyuzhuo <yuzhuo_liu@realsil.com.cn>

ZhiyuanTang17 committed:
drivers: entropy: support trng driver for rtl87x2g series

ZhiyuanTang17 committed:
drivers: flash-controller: support flash driver for rtl87x2g series

YuzhuoLiuRTK committed:
dts: bindings: relocate realtek,rtl87x2g-usb.yaml

Signed-off-by: liuyuzhuo <yuzhuo_liu@realsil.com.cn>

YuzhuoLiuRTK committed:
dts: bindings: adc: modify rtl87x2g binding file

Signed-off-by: liuyuzhuo <yuzhuo_liu@realsil.com.cn>

YuzhuoLiuRTK committed:
dts: bindings: kscan: modify rtl87x2g binding file

Signed-off-by: liuyuzhuo <yuzhuo_liu@realsil.com.cn>

YuzhuoLiuRTK committed:
dts: bindings: pinctrl: modify rtl87x2g binding file

Signed-off-by: liuyuzhuo <yuzhuo_liu@realsil.com.cn>

ZhiyuanTang17 committed:
soc: rtl87x2g: correct the CONFIG_NUM_IRQS

Max number of IRQn_Type in bee4 is 108, so CONFIG_NUM_IRQS should be 109. If CONFIG_NUM_IRQS is set to a value greater than 130, it can cause the scb->ccr register to be unintentionally cleared, leading to the phenomenon of cache being disabled.

ZhiyuanTang17 committed:
soc: rtl87x2g: move kernel and arch code to ram

YuzhuoLiuRTK committed:
drivers: kscan: change scan_clk for rtl87x2g

Update scan_clk to 312.5 kHz for faster scan speed.

Signed-off-by: liuyuzhuo <yuzhuo_liu@realsil.com.cn>
